ProjFS.Mac/PrjFSKext/PrjFSKext/PrjFSProviderUserClient.cpp

```diff
@@ -80,6 +80,16 @@ bool PrjFSProviderUserClient::initWithTask(
     return false;
 }
 
+void PrjFSProviderUserClient::stop(IOService* provider)
```

Comment:
I noticed a similar comment in `PrjFSLogUserClient::cleanUp()`, and it looks like in `PrjFSLogUserClient::cleanUp()` is called from `PrjFSLogUserClient::free()`.

Are there advantages to doing this cleanup in `stop(...)` instead of `free()` (or was using `free()` not an option here)?

Reply:
`free()` is effectively a destructor in the `OSObject` system, so that's conceptually more a place to free resources. For `IOService` subclasses (including `IOUserClient`) you've got this lifecycle:

 * Object creation with `new`
   * Object initialisation with `init…()`
     * Attaching service to provider object in IORegistry graph using `attach()`
       * Service start with `start()`
         * (useful life of the object)
         * Service termination request (`terminate()`)
       * Service shutdown with `stop()`
     * Detaching service from provider (`detach()`)
 * Object eventually hits retain count of 0 (hopefully), which causes `free()` to be called, and subsequent destruction/deallocation with `delete`

If any of `init`, `attach`, or `start` fail, it falls back out to the previous level of indentation in the above hierarchy. So `stop()` is only called if `start()` succeeded, etc. Similarly, `free()` will only be called once retain count hits 0, which could conceivably be a lot later than `stop()`. (Or never if there's a retain leak.)

So `stop()` is probably a better place for this, and if anything, the `KextLog_DeregisterUserClient()` call in `PrjFSLogUserClient` should be moved to its `stop()` function. (That one is less serious if it doesn't get called though.)

Reply:
Ok thanks for the details! Based on your description I agree `stop()` is the better place.